jax/experimental/optix.py

```diff
@@ -0,0 +1,361 @@
+
+An (optional) `chainer` utility can be used to build custom optimizers by
+chaining arbitrary sequences of transformations. For any sequence of
+transformations `chainer` returns a single `init_fn` and `update_fn`.
+
+An (optional) `apply_updates` function can be used to eventually apply the
+transformed gradients to the set of parameters of interest.
+
+Separating gradient transformations from the parameter update allows to flexibly
+chain a sequence of transformations of the same gradients, as well as combine
+multiple updates to the same parameters (e.g. in multi-task settings where the
+different tasks may benefit from different sets of gradient transformations).
+
+Many popular optimizers can be implemented using `optix` as one-liners, and,
+for convenience, we provide aliases for some of the most popular ones.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+from jax import numpy as jnp
+from jax.tree_util import tree_leaves
+from jax.tree_util import tree_multimap
+
+
+### Composable gradient transformations. ###
+
+
+ClipState = collections.namedtuple("ClipState", "")
+
+
+def clip(max_delta):
+  """Clip updates element-wise.
+
+  Args:
+    max_delta: the maximum size of an update, for each variable
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(_):
+    return ClipState()
+
+  def update_fn(updates, state):
+    updates = tree_multimap(
+        lambda g: jnp.clip_by_value(g, -max_delta, max_delta), updates)
+    return updates, state
+
+  return init_fn, update_fn
+
+
+ClipByGlobalNormState = collections.namedtuple("ClipByGlobalNormState", "")
+
+
+def _global_norm(items):
+  return jnp.sqrt(jnp.sum([jnp.sum(x**2) for x in tree_leaves(items)]))
+
+
+def clip_by_global_norm(max_norm):
+  """Clip updates using their global norm.
+
+  References:
+    [Pascanu et al, 2012](https://arxiv.org/abs/1211.5063)
+
+  Args:
+    max_norm: the maximum global norm for an update.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(_):
+    return ClipByGlobalNormState()
+
+  def update_fn(updates, state):
+    g_norm = _global_norm(updates)
+    trigger = g_norm < max_norm
+    updates = tree_multimap(
+        lambda t: jnp.where(trigger, t, t * (max_norm / g_norm)), updates)
+    return updates, state
+
+  return init_fn, update_fn
+
+
+TraceState = collections.namedtuple("TraceState", "trace")
+
+
+def trace(decay, nesterov):
+  """Compute a trace of past updates.
+
+  Args:
+    decay: the decay rate for the tracing of past updates.
+    nesterov: whether to use nesterov momentum.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(params):
+    return TraceState(trace=tree_multimap(jnp.zeros_like, params))
+
+  def update_fn(updates, state):
+    f = lambda g, t: g + decay * t
+    update_trace = tree_multimap(f, updates, state.trace)
```

Comment:
Almost every optimizer `update_fn` (all of them except `clip_by_global_norm`) performs `tree_multimap` over an inner per-parameter update. Is that a pattern that you can extract out like `@optimizer` in optimizers.py or is there a reason why that doesn't work?

Reply:
I purposefully diverged from `optimizers.py` in this aspects for two reasons:
1) there are several gradient transformations of interest that require to consider the entire gradient and cannot be computer variable by variable. The `clip_by_global_norm` transformation included here is one example, but others could be PopArt, KFac, etc...
2) it really is just a handful of characters that are saved, and at the cost of introducing an additional level of indirection and a reduction in flexibility.